geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/process/JoinProcess.scala

```diff
@@ -0,0 +1,121 @@
+              unique: SimpleFeatureCollection,
+              @DescribeParameter(name = "join", description = "Joined features")
+              join: SimpleFeatureCollection,
+              @DescribeParameter(name = "attribute", description = "Attribute field to join on", min = 1)
+              attribute: String,
+              @DescribeParameter(name = "bins", description = "Return BIN records or regular records", min = 0)
+              bins: java.lang.Boolean,
+              @DescribeParameter(name = "binDtg", description = "Date field to use for BIN records", min = 0)
+              binDtg: String,
+              @DescribeParameter(name = "binTrackId", description = "Track field to use for BIN records", min = 0)
+              binTrackId: String,
+              @DescribeParameter(name = "binLabel", description = "Label field to use for BIN records", min = 0)
+              binLabel: String,
+              monitor: ProgressListener): SimpleFeatureCollection = {
+
+    import org.locationtech.geomesa.filter.ff
+
+    import scala.collection.JavaConversions._
+
+    logger.trace(s"Attempting join query on ${join.getClass.getName}")
+
+    if (join.isInstanceOf[ReTypingFeatureCollection]) {
+      logger.warn("WARNING: layer name in geoserver must match feature type name in geomesa")
+    }
+
+    val descriptor = join.getSchema.getDescriptor(attribute)
+    require(descriptor != null, s"Attribute $attribute does not exist in the joined feature collection")
+
+    require(unique.getSchema.getTypeName == UniqueProcess.SftName &&
+        unique.getSchema.getDescriptor(0).getLocalName == UniqueProcess.AttributeValue,
+      s"Input feature collection is not the result of a unique process. Expected " +
+          s"${SimpleFeatureTypes.encodeType(UniqueProcess.createUniqueSft(descriptor.getType.getBinding, histogram = false))} " +
+          s"but got ${SimpleFeatureTypes.encodeType(unique.getSchema)}"
+    )
+
+    val attributeProperty = ff.property(attribute)
+    val joinFilters = SelfClosingIterator(unique).map { sf =>
+      ff.equals(attributeProperty, ff.literal(sf.getAttribute(0)))
+    }
+
+    val binOptions = if (bins != null && bins) {
+      import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+      val sft = join.getSchema
+      val dtg = Option(binDtg).orElse(sft.getDtgField).getOrElse {
+        throw new IllegalArgumentException("Please specify binDtg for BIN output format")
+      }
+      val track = Option(binTrackId).orElse(sft.getBinTrackId)
+      Some(EncodingOptions(dtg, track, Option(binLabel)))
+    } else {
+      None
+    }
+
+    val result = if (joinFilters.isEmpty) {
+      new DefaultFeatureCollection(null, join.getSchema)
+    } else {
+      val filter = ff.or(joinFilters.toList)
+      val visitor = new QueryVisitor(join, filter)
+      join.accepts(visitor, new NullProgressListener)
```

Comment:
the query visitor gets intercepted by our data store, so it results in a new geomesa query and doesn't run directly against each feature. But yeah, in general you would end up with a massive OR query. In general we handle those pretty well, but the onus is on the user to make suitable queries.
The impetus for this functionality is something like a summary feature combined with multiple detail features. We look up the summary by ID (so only a single value comes back), then join against the detailed features. Instead of making it a 2-stage query handled by the client, it's all handled server side in this WPS.